Hand of God: Resurgence

RemieRichards opened the pull request:
## HAND OF GOD GAMEMODE

AI+Cult+Gang with some more RTS elements. There are two gods, tasked with certain objectives: Growing their cults, Having their cults build many shrines, Sacrificing enemy prophets and finally, KILLING THE OTHER GOD!
The god gains "Faith" based on the number of cultists it has and the number of powerpylons it's cult has built for it.
The god can place down "construction sites" that require metal/glass/(lesser/greater) gems to finish construction
## Code changes:
- Rewrote antag preference saving to not be a bitfield anymore, allowing for more than 16 possible antag types (functionally infinite now), This includes a proc on preferences to update out of date savefiles, players should never notice.
- `transfer_to()` no longer requires a `/living` mob
- Antag Huds may now use a `/living` mob in some cases, I tried to make it 100% compatible but it's a fucking mess and it's pissing me off because it totally should work but just doesn't
- Minds are now initialised on all `/mob` that have a login() (Except ghosts, Ghosts were pretty much the only reason minds were limited to `/mob/living`, so that ghosts don't lose the reference to their human (for objectives and junk), so existing functionality is maintained) 
- `/gun_turret`s now have a `base` var, where all range and shooting checks are done from, to allow turrets inside other objects, to allow for more functionality without snowflake.
## Progress map:
- [x] Rewrote Antag Preference saving to use a list, allowing functionally infinite antag types (Or atleast more than the current 16 antag limit)
- [x] Gamemode Framework
- [x] God mob
- [x] God Powers
- [x] Structures
- [x] Traps
- [x] Objectives
- [x] Antag Hud - Sort of
- [x] Set up Icons/Mock up missing Icons
## Singleplayer Testing Phase - <B>100%</B>
- [x] Compilation
- [x] Merge Conflicts Solved
- [x] Fix antag huds not working on `/mob` - Good enough
- [x] Fix `transfer_to()` not working on `/mob`
- [x] Fix runtimes - Tied to antag huds
- [x] Fix construction
- [x] Fix preferences screen not updating
## Feedback Based Updates Phase - <B>100%</B>
- [x] Split tasks up between people
  http://piratepad.net/FT9RnWLF3C - <B>tkdrg & Remie</B>
- [x] MERGE CONFLICTS - <B>tkdrg</B>
- [x] Healing fountain has no construction sprite - <B>tkdrg</B>
- [x] fixes "If a god uses an ability with no target, it still uses up faith." - <B>tkdrg</B>
- [x] Nexii failing to delete - <B>tkdrg</B>
- [x] Convert messages - <B>tkdrg</B>
- [x] Informing gods to place their nexus in a safe spot - <B>tkdrg</B>
- [x] Round end message - <B>tkdrg</B>
- [x] Admin check antagonist panel shows Gods/Followers - <B>tkdrg</B>
- [x] Typos - <B>tkdrg</B>
- [x] Building click delays - <B>tkdrg</B>
- [x] Defence Pylon ON/OFF - <B>tkdrg</b>
- [x] Holy Water/Implant deconversion - <B>tkdrg</b>
- [x] No followers = Death - <B>tkdrg</b>
- [x] Prophet2God Communication - <B>Remie</B>
- [x] Gods can now cancel their structure construction sites - <B>Remie</B>
- [x] New structure construction/trap placement UI - <B>Remie</B>
- [x] Reduce structure costs by 5/10 across the board - <B>Remie</B>
- [x] A way for teams to spot the enemy Prophet, for objectives and loot - <B>Remie</B>
- [x] Being able to convert a structure to your team's side - <B>Remie</B>
- [x] Converted turrets immediately recognise your side as their own (No teamkilling) - <B>Remie</B>
- [x] Removes holyslumber for being OP and kind of bullpoop - <B>Remie</B>
## Credits
- <B>Neerti</B> - Idea, Concept, Original code, Sprites
- <B>WJ</B> - Sprites
- <B>Remie</B> - Code, Sprite edits
- <B>tkdrg</B> - Code, Compiling list of feedback requests: http://piratepad.net/FT9RnWLF3C

RemieRichards committed:
Hand of god gamemode!

be_special and antag defines are no longer limited to 16 values (1 - 32768) by being changed from a bitfield to a list, an appropriate block of code exists to automatically update bitfield versions to the new list format. the list format allows us to have (functionally) infinite antag types.

Razharas commented:
So a gang/cult/AI combination huh
I heard ideas like this before, this can be nice

RemieRichards commented on code/modules/client/preferences_toggles.dm:
```diff
@@ -214,7 +214,10 @@
 	var/role_flag = be_special_flags[role]
 	if(!role_flag)	return
-	prefs.be_special ^= role_flag
+	if(role_flag in prefs.be_special)
```
REQUESTED CHANGE (from Oranges and MSO):
Change this to not use a basic byond list anymore, so that we get a *nicer* UI that actually tells us which roles are enabled and which are disabled.

Then perhaps in a later PR allow preferences to be opened anytime so players get the same editing window+UI as in the lobby, you know, the not shit one.

GunHog commented:
Yes. Yes. This needs to exist. PLEASE LET IT EXIST. START A BLOODY KICKSTARTER FOR RR IF IT HELPS HER GET IT DONE! :+1:

Lumipharon commented:
Fuck yes. I was always pretty sad that this never got finished.
Hopefully it does this time around.

WJohn commented:
Oh my god. IT'S BACK.

REPLACE CULT WITH THIS SHIT.

Fox-McCloud commented:
@WJohn 

At least give the cult refactor a chance, no one has even played it yet, as far as I know =p

GunHog commented:
Because you can make PRs to other PRs, AnturK and Xhuis could potentially work together with you to complete this project!

RemieRichards commented:
Not necessary, I've left a mess of code only I really understand at the moment, once the initial structure is complete they are however, free to PR new things that fit the Hand of God theme, be it additions/removals or tweaks. That should be in about 2 days

@Xhuis @AnturK

AnturK commented:
Yeah, probably best idea to wait until core code is stable

Steelpoint commented:
Man this game mode has been a thing forever, didn't Neerti start work on it a year ago?

Well good luck either way, its a unique game mode concept.

demonfiren commented:
Is this the Masks of Nar-Sie that Bay wanted to do years ago but never happened?

If so, FUCK YES.

JJRcop commented:
For reference: https://tgstation13.org/phpBB/viewtopic.php?f=16&t=115

ColonicAcid commented:
I hope one of you three have atleast played Dominions. Neerti borrowed heavily from it, that's what made it interesting. Please don't ruin it.

Supermichael777 commented:
suggestion: make its so that cultists prayers go directly to their god. have this be the only thing gods can hear that isn't in their vision.

Madventurer commented:
Cool, I suggested this somewhere over two years ago.
Great that someone finally dug it up or just made the same idea.
http://forums.yogstation.net/index.php?threads/news-gamemode-two-powers.648/

RemieRichards commented:
[Soon](http://i.imgur.com/UqTYIm4.png)

Sooooooon.

Fox-McCloud commented:
@RemieRichards 

Is there an explanation, anywhere, of what this game-mode entails, exactly?

RemieRichards commented:
@Fox-McCloud Dig around for Neerti's old documents, it's largely unchanged.
This one for example https://tgstation13.org/phpBB/viewtopic.php?f=16&t=115

Fox-McCloud commented:
@RemieRichards 

Thank you! Seems like a very very interesting game mode that has a lot of potential.

JJRcop commented:
@Fox-McCloud Well it's not like I posted the exact same link or anything that you didn't pay attention to ;_;

Fox-McCloud commented:
@JJRcop 

Apologies; I overlooked it.

RemieRichards committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into HandOfGod-Resurgence

Conflicts:
	code/__HELPERS/unsorted.dm
	code/game/objects/items/stacks/sheets/sheet_types.dm
	code/modules/client/preferences.dm
	icons/mob/feet.dmi
	icons/mob/head.dmi
	icons/mob/inhands/items_lefthand.dmi
	icons/mob/inhands/items_righthand.dmi
	icons/mob/mob.dmi
	icons/obj/clothing/hats.dmi
	icons/obj/clothing/shoes.dmi
	icons/obj/clothing/suits.dmi
	icons/obj/items.dmi
	icons/obj/projectiles.dmi

RemieRichards committed:
Fixes botched merge on DME

Razharas commented:
Hmm feels like a nice time to make antag datums

Dorsisdwarf commented:
Is this being tested?

RemieRichards commented:
@Dorsisdwarf 

>Singleplayer Testing Phase - 85%

Bawhoppen commented:
is this going to mystically fix all of /tg/'s problems?

Dicks1 commented:
How many players minimum do you think this will need?

RemieRichards commented:
I'd say atleast 8 (4 per team) onwards.

Steelpoint commented:
One god, one prophet and two underlings seems good.

RemieRichards commented:
@Steelpoint well, it'd be One god and 3 underlings, the prophet would be chosen by the God, not the game.

Razharas commented:
@RemieRichards

RemieRichards commented:
@Razharas no shit there's merge conflicts, you just caused em.

Razharas commented:
I din do nuffin

Razharas commented:
Please dont abandon this
@RemieRichards

RemieRichards commented:
1. it's practically finished
2. I said by the end of this week, it's only Wednesday night

RemieRichards committed:
About a million changes, but largely: Functional (Sort of) Antag Huds, Mind fixes, and Final glue to get the mode working.

RemieRichards committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into HandOfGod-Resurgence

Conflicts:
	code/modules/mob/living/simple_animal/guardian/guardian.dm
	icons/mob/head.dmi
	icons/mob/suit.dmi
	icons/obj/clothing/hats.dmi
	icons/obj/clothing/suits.dmi
	icons/obj/storage.dmi

RemieRichards commented:
## IT'S THE END OF THE WORLD AS WE KNOW IT

this can be merged/pulled for Testing.

RemieRichards commented:
Fix for travis upcomming, dammit it @WJohn and @GunHog

RemieRichards committed:
Fixes compile issue relating to the Royal Xeno PR.

Chronitonity commented:
They dont pay you enough

GunHog commented:
*Gun Hog silently whimpers in the corner.*

Neerti commented:
:+1:

optimumtact commented:
byond://singulo.io:45000

RemieRichards commented:
## Just going to do some Polish tomorrow, and it should be good, DNMing just for a day

Razharas commented:
Less items - more abilities for god to help in battle outside the conduit range
Then it would be real good

as334 commented:
For Polish it'd be nice if there was a way to turn of defensive turrets as those things are really janky, including shooting your own team and killing people you're trying to drag to your convert spot. Also Prohpet REALLY needs a way to communicate with their god.

Durkel commented:
We're doing a playtest currently, I'll list off some bugs I've seen so far or heard.

1)Buildings have no click delay
2) Shock traps trigger on friendly cultists (intended?)

astralenigma commented:
Gods spawning without a prophet is supposed what is not supposed is a god spawning without followers.
Also there should be an option for god to convert some random person when all followers are dead, otherwise the god is effectively dead.

Razharas commented:
1 - No list of followers in the check antags panel
2 - No way of making someone a god, no way of making more than two gods

astralenigma commented:
2 is kinda there by default since the gods are coloured base and there's only red and blue, to add more gods you would need to add more colours or the gods would control the same side.

Steelpoint commented:
Feedback thread here: https://tgstation13.org/phpBB/viewtopic.php?f=10&t=4869

Jordie0608 commented on data/mode.txt:
```diff
@@ -1 +1 @@
-extended
\ No newline at end of file
+handofgod
```
:whale2:

Razharas commented:
@RemieRichards 
FINISH HIM!!
Mergeconflictality!

Bluespace commented:
Dream is dead

GunHog commented:
Ladies and Gentlemen, Dismemberment, again!

Shadowlight213 commented:
Rip

Razharas commented:
12 days until the doom

Razharas commented:
@RemieRichards

RemieRichards commented:
## <B> Feedback Based Updates Phase - 0%</B>

DUN DUN DUN!?

PKPenguin321 commented:
happening?

Anonus commented:
DUN DUN DUN!

tkdrg committed:
Merge remote-tracking branch 'upstream/master' into hog-mergeconflicts

Conflicts:
	code/__DEFINES/hud.dm
	code/__HELPERS/game.dm
	code/_onclick/hud/_defines.dm
	code/_onclick/hud/hud.dm
	code/game/gamemodes/antag_spawner.dm
	code/game/gamemodes/game_mode.dm
	code/game/machinery/turrets.dm
	code/modules/client/preferences_toggles.dm
	code/modules/events/alien_infestation.dm
	code/modules/mob/dead/observer/observer.dm
	code/modules/mob/living/carbon/alien/special/alien_embryo.dm
	code/modules/mob/living/simple_animal/revenant/revenant.dm
	icons/mob/back.dmi
	icons/mob/head.dmi
	icons/mob/hud.dmi
	icons/mob/inhands/items_lefthand.dmi
	icons/mob/inhands/items_righthand.dmi
	icons/mob/mob.dmi
	icons/mob/suit.dmi
	icons/obj/clothing/hats.dmi
	icons/obj/items.dmi
	icons/obj/projectiles.dmi
	icons/obj/storage.dmi

RemieRichards committed:
Merge pull request #13 from tkdrg/hog-mergeconflicts

Hog mergeconflicts

RemieRichards commented:
## Feedback Based Updates Phase - 35%
@tkdrg has been hammering away on his tasks!

tkdrg committed:
Hand of God Fixes/Tweaks

Fixes healing fountain construction sprite.
Nexus are now destroyed properly.
Borged cultists will now lose antag status properly.
Round end message fixed.
Feedback messages improved to be clearer, and lots of typos fixed.
Smite no longer uses up faith if there is no valid target in the turf.
Adds an atom/attack_god() proc.
Fixes the greeting messages for prophets and followers, improves god msg.
Fixes nexus being spawned on dense turf after 2min, and general sanity.
Fixes attacking structures with abstract items.
Adds click delay and attack animation for attacking structures.
Fixes a lot more typos.
Fixes defense pylon friendly fire.
Makes defense pylon togglable on its god's click.
Cleans up gun_turret code a little in general.
Adds deconversion by loyalty implant or holy water (basically rev+cult).
Gods will now die when they have no followers left.
Adds the HoG gods to the traitor panel.

RemieRichards commented:
## Feedback Based Updates Phase - 55%
Me and @tkdrg will be busy this week, but progress is looking good!

Bluespace commented:
REFUEL THE HYPETRAIN!

Supermichael777 commented:
Team colored pope hats?
![swag pope hats](https://dtsft.files.wordpress.com/2013/02/the-pope-and-hat-1.jpg)

RemieRichards commented:
@Supermichael777 yep, it's in the task document as a solution to making Prophets obvious to the other side, a Pope outfit with some decent stats and stuffs.

RemieRichards committed:
Merge pull request #14 from tkdrg/groundHoGday

Hand of God Fixes/Tweaks

tkdrg committed:
Merge remote-tracking branch 'upstream/master' into fuk

Conflicts:
	data/mode.txt
	icons/mob/mob.dmi
	icons/mob/suit.dmi
	icons/obj/clothing/gloves.dmi
	icons/obj/clothing/hats.dmi
	icons/obj/clothing/shoes.dmi
	icons/obj/clothing/suits.dmi

RemieRichards committed:
Merge pull request #15 from tkdrg/fuk

Fuk

RemieRichards commented:
@tkdrg is the real mvp.

RemieRichards committed:
More Hand of God updates:
* Prophet2God communication
* Cancel structure construction
* New Construction/Trap UI
* Reduced structure costs
* Structure conversion
* Prophet gear
* Removes holy slumber for being OP

RemieRichards commented:
## Feedback Based Updates Phase - <B>100%</B>

WHAT COULD IT MEAN!?!

Cheridan commented on data/mode.txt:
```diff
@@ -1 +1 @@
-extended
+handofgod
```
dork

Iamgoofball replied:
:+1:

RemieRichards replied:
we dont use this config file on the server :^)

RemieRichards committed:
Fixes default gamemode back to extended, this was changed by the "Change Gamemode Verb" when I was testing.

PKPenguin321 commented:
oh goodness
